### PR TITLE
Fix comment operators in c-mode

### PR DESCRIPTION
**Pull request by rcj**
Don't add space before a line comment if it is at the indentation of
the current line (single line comment).

Add a space before a line comment if it is not at the indentation of
the current line.

This requires to keep track of the actual insertion position (ignoring
leading white-space) and passing it to the operator. This might also
be useful for enhancing other c-mode operators by querying cc-mode for
the syntactic symbols at the insertion point.

### CODE REVIEW

File: electric-operator.el [hunk 1]
```diff
@@ -201,9 +201,15 @@ Whitespace before the operator is captured for possible use later.
...
        (car it)))
 
-(defun eval-action (action point)
+(defun eval-action (action point state)
   (cond
-   ((functionp action) (save-excursion (goto-char point) (funcall action)))
+   ((functionp action)
+    (save-excursion
+      (goto-char point)
+      ;; For backwards-compatibility only pass state to
```
Comment (rcj) [inline electric-operator.el]: Not sure if there is a better way of doing this without breaking backwards-compatibility

File: features/c-mode-basic-operators.feature [hunk 2]
```diff
@@ -189,27 +204,26 @@ Feature: C basic operators
     When I type "result = foo * *bar"
     Then I should see "result = foo * *bar"
 
-  # # This doesn't actually test what it should, it always passes. Possible
-  # # ecukes bug?
```
Comment (rcj) [inline features/c-mode-basic-operators.feature]: This looks indeed like an ecukes bug. Changing (call-interactively #'move-end-of-line) to (funcall #'move-end-of-line nil) fixes this (see electric-spacing-steps.el), however, I currently don't understand why. I created a similar espuds ert test, but there it is working fine. Not sure if the py-string syntax does anything fancy that leads to this. I used plain (insert) in the ert test and didn't look at the py-string implementation.

**Comment by davidshepherd7** [comment]
The test failures seem to be because of something on Travis causing `cask install` to fail. Not sure what but don't worry about it for now.

File: electric-operator.el (hunk 1 above)
Comment (davidshepherd7) [reply electric-operator.el]: Stefan Monnier [thinks trying calling it both ways and catching the exception is the way to go](https://emacs.stackexchange.com/a/29247/237), but there doesn't seem to be a *good* way to handle this which is a shame. His problem with this version is that it won't handle some more advanced features like advised functions.

I think this is fine for now (unless you want to change it). I'll tweak it in the future if it causes problems.

File: electric-operator.el [hunk 3]
```diff
@@ -225,7 +231,9 @@ Whitespace before the operator is captured for possible use later.
       (let* ((pre-whitespace (match-string 1))
              (op-match-beginning (match-beginning 0))
              (op-match-end (match-end 0))
-             (spaced-string (eval-action action op-match-beginning)))
+             (op-start-pos (+ op-match-beginning (length pre-whitespace)))
+             (state `((start-pos . ,op-start-pos)))
```
Comment (davidshepherd7) [inline electric-operator.el]: As much as I dislike how verbose emacs' defstruct stuff is", I think this should probably be one so that people adding functions can have a nicer interface than `caar` etc.

File: electric-operator.el (hunk 1 above)
Comment (davidshepherd7) [reply electric-operator.el]: I found that we can do `((looking-back-locally "^\\s-*") "/")` instead of checking if the initial operator insertion point is at the indentation. It passes all the tests, and I can't think of anywhere that it won't work. So to avoid the extra complexity of having variable numbers of arguments I changed it to do that.

Can you think of any edge cases where that won't work?

**Comment by davidshepherd7** [comment]
I've merged it with a few tweaks, thanks :+1:

File: electric-operator.el (hunk 1 above)
Comment (rcj) [reply electric-operator.el]: I had a reason yesterday which is why I replaced my initial version which was using `looking-back-locally` with `state`, however, I currently cannot remember. So most likely some unusual edge case.

Comment (davidshepherd7) [reply electric-operator.el]: Ah ok, well if you run into the issue again we can try to fix it.

One weird thing I found while messing with the tests is that `c-electric-slash` etc actually modify the indentation when you type the second `/` (unless you've turned it off). Maybe that was causing some odd edge cases?

Comment (rcj) [reply electric-operator.el]: I don't have c-electric-* turned on, so that shouldn't have been it. I'll take a note next time I come across it/something similar.
